libhif/hif-source.c

```diff
@@ -1634,33 +1634,63 @@ hif_source_copy_package (HyPackage pkg,
 			    hif_source_copy_progress_cb, state, error);
 }
 
-struct DownloadState {
-	HifState *state;
+typedef struct
+{
 	gchar *last_mirror_url;
 	gchar *last_mirror_failure_message;
-};
+} MirrorFailureData;
+
+typedef struct
+{
+	HyPackage pkg;
+	HifState *state;
+	MirrorFailureData *mirror_failure_data;
+} PackageDownloadData;
 
 static int
 package_download_update_state_cb (void *user_data,
 				  gdouble total_to_download,
 				  gdouble now_downloaded)
 {
-	struct DownloadState *dlstate = user_data;
-	return hif_source_update_state_cb (dlstate->state, total_to_download, now_downloaded);
+	PackageDownloadData *data = user_data;
+	HifState *state_child;
+
+	state_child = hif_state_get_child (data->state);
```

Comment:
When I looked at this patch my question was "how does it fix the issues that blocked my earlier attempt", which was mainly around progress reporting.

This looks like it's introducing a new child state every time we get a progress update, which is wrong.  I think we need to create a child state for each package, stick them in a hash table, then look up the child state and update it.  Or something like that.  Right?